### PR TITLE
gh-96859: [argparse] Avoid O(N^2) behavior while consuming optionals

### PR DESCRIPTION
**Pull request by alexeypa**
The loop consuming optional and positional arguments in ArgumentParser._parse_known_args() rescans option_string_indices on every iteration of the loop in order to compute the next option index:

```py
next_option_string_index = min([
    index
    for index in option_string_indices
    if index >= start_index])
```

This becomes very slow when parsing a very long list of arguments, e.g. when parsing a response file.

This commit refactors the loop to scan the option indices list only once.

* Issue: gh-96859

### CODE REVIEW

**Comment by rahul-kumi** [comment]
@bitdancer, please share your reviews.

File: Lib/argparse.py
Comment (DanielNoord) [inline Lib/argparse.py]: This also needs to be updated.

Is `option_tuple` even used? I'm not sure what the standard is for these situations but you could also consider just passing in the values directly.

File: Lib/argparse.py
Comment (DanielNoord) [inline Lib/argparse.py]: This comment should be updated to reflect no longer looking only for indices.

File: Lib/argparse.py
Comment (DanielNoord) [inline Lib/argparse.py:2071]: In the old code. This passes L2084.

In this particular case will L2098 then always be true? If so, should we also re-assign `start_index`.
If not I think this code path would go on until L2013 without any other `continues`. Is this a change in behaviour? Or is this simply never not true?

**Review by DanielNoord — CHANGES_REQUESTED** [review]
Code itself looks good, although I had some trouble trying to understand the new flow based on the old flow. _Having no particular experience with the code didn't help._

I had one question and some comments about comments. Those should be updated a little bit.

File: Lib/argparse.py
Comment (alexeypa) [reply Lib/argparse.py:2071]: Thanks for taking a look!

I think the old and new loop are fully equivalent. We can only hit L2075 in the new code after either `consume_positionals()` or `consume_optional()` extend `start_index` past the next `option_tuple_index`. In such a case, we will effectively skip any `option_tuple_index` smaller than `start_index` which is exactly the same behavior the old code achieved with `next_option_string_index = min([index for index in option_string_indices if index >= start_index])`.

Looking at the old code, step by step.

```
            next_option_string_index = min([
                index
                for index in option_string_indices
                if index >= start_index])
```

^^^ this select `next_option_string_index` greater or equal to `start_index`. This means that the condition `if start_index != next_option_string_index:` (L2084, old) can be rewritten as `if start_index < option_tuple_index:` (L2078, new).

Handling the value returned by `consume_positionals()` is tricky. The old loop could restart the iteration with the same value of `next_option_string_index`. The new loop has to make forward progress, so this is handled differently. There are four possible cases:

1. The `consume_positionals()` return the same index that was passed in. We will fall through to L2097 which is the same logic as the new L2088.
2. `consume_positionals()` increments the returned index but it is still less than `next_option_string_index`/`option_tuple_index`. The old code would loop (L2091) and find the same `next_option_string_index`. This time `consume_positionals()` will not advance the index, so we will fall through to L2097. The new code handles this by falling to L2088 without restarting the iteration.
3. The index returned by `consume_positionals()` matches `positionals_end_index`/`option_tuple_index`. Again the old code would loop but ultimately get to L2097 and skip the condition. The new code goes directly to L2088.
4. The index returned by `consume_positionals()` is greater than `positionals_end_index`/`option_tuple_index`. Both old and new code would loop and skip at least one optional tuple.

Does this answer your question?

File: Lib/argparse.py
Comment (alexeypa) [reply Lib/argparse.py]: I updated the comment at L1936.

File: Lib/argparse.py
Comment (alexeypa) [reply Lib/argparse.py]: Unpacking `option_tuple` seems like a good alternative.

**Review by alexeypa — COMMENTED** [review]
BTW, what is the etiquette around commits addressing review feedback that are not really meaningful changes by themselves? Should I squash them before the PR is merged?

File: Lib/argparse.py
Comment (DanielNoord) [reply Lib/argparse.py:2071]: Thanks for the clear explanation!

One follow-up. My initial point was about case 4. In that case, does old L2090 not reassign `start_index`? The continue on new L2071 doesn't reassign, which is why I thought there might be a subtle difference.

File: Lib/argparse.py
Comment (DanielNoord) [reply Lib/argparse.py]: I still think that L1921 should also be updated. The comment refers to `indices` which was the name of the old variable.

File: Lib/argparse.py
Comment (alexeypa) [reply Lib/argparse.py:2071]: > My initial point was about case 4. In that case, does old L2090 not reassign start_index? The continue on new L2071 doesn't reassign, which is why I thought there might be a subtle difference.

The new code reassigns `start_index` on L2075. Basically the old and new code maintain the same invariant - `start_index` will be set to the value returned by `consume_positionals()`. It is just the old code passes the same value through `positionals_end_index` first.

The continue at L2071 does not need to reassign `start_index` because at this point the code just tries to rewind `option_tuples` to the next index that is `>= start_index`.

File: Lib/argparse.py
Comment (alexeypa) [reply Lib/argparse.py]: Done.

File: Lib/argparse.py
Comment (DanielNoord) [reply Lib/argparse.py:2071]: Ah yeah. Wow, for some reason I couldn't wrap my around that. Sorry for making such a point about this, I should have understood this. Thanks for the careful explanation anyway!

**Review by DanielNoord — APPROVED** [review]

**Comment by alexeypa** [comment]
> Tests / Windows (x86) (pull_request) Failing after 31m 
> https://github.com/python/cpython/actions/runs/3091293898/jobs/5001192381

This does not seem related to the changes on the PR. Also the latest commit only touched comments.

**Comment by AlexWaygood** [comment]
> > Tests / Windows (x86) (pull_request) Failing after 31m
> > https://github.com/python/cpython/actions/runs/3091293898/jobs/5001192381
> 
> This does not seem related to the changes on the PR. Also the latest commit only touched comments.

Try closing and reopening the PR to get the check to re-run.

**Comment by AlexWaygood** [comment]
> BTW, what is the etiquette around commits addressing review feedback that are not really meaningful changes by themselves? Should I squash them before the PR is merged?

Please don't! All CPython PRs are squash-merged anyway, so you don't need to; and PR authors squashing commits together makes it harder for reviewers to see what changed between commits due to the GitHub UI :)

**Comment by alexeypa** [comment]
A friendly ping to bring some attention to this PR.

@rhettinger, would you be the best person to do a core review? You listed as the maintainer of argparse.py at the top of the file. Please let me know if I should ping someone else.

**Comment by hpaulj** [comment]
I used to contribute a lot to argparse as paul.j3, but haven't followed it much since the github move.  So this will be a general comment.

Backward compatibility has been a major issue; I can think of several changes that either had to be removed, or have remained and continue to cause problem.  Thus I favor bug fixes, and feature enhancements that have little chance of causing problems.  I discourage core changes that aren't essential.

This looks like a core change simply for the sake of performance.  I'd vote against it, unless it has been thoroughly tested, not just read through by one or two people.  I don't think there are enough argparse developers to do that well.  

Most users don't use argparse for large inputs.  If performance came up on StackOverFlow, I'd tell them to read the input(s) from a file, and reserve argparse for short lists - file names and controls, not data.

**Comment by alexeypa** [comment]
Thanks for your feedback, @hpaulj. The change is indeed fixes a performance issue and attempts to not modify the existing behavior in any way. The test_argparse.py suite looks pretty comprehensive and tests a number is of cases involving append actions and a mix of optional/positional arguments. However it does not test the mix of argparse and 3rd party apps using it - it is not an appcompat test suite. I don't believe I can provide any evidence showing that this change does not introduce breakage with in a large body of 3rd party code. If this means this change does not meet the bar - so be it.

>  If performance came up on StackOverFlow, I'd tell them to read the input(s) from a file, and reserve argparse for short lists - file names and controls, not data.

The specific case this issue came up in was parsing a response file generated by a build system. It was a reasonable legitimate case of passing diverse command line parameters via a file due to size constraints. I'd agree this is a corner case.

**Comment by hpaulj** [comment]
From a quick read of your changes it looks like you are replacing the set, `option_string_indices` with list `option_tuples` and `option_tuple_index`.

Years ago I added `parse_intermixed_args`.  I created a whole new pair of functions rather than add some sort of parameter and alternative parsing path in the default `parse_args`.   You might consider something similar with a `parse_faster_args`.   At least until we get more feedback, leave the current parser as the default, but provide interested users with an alternative.  

A down side to that is it won't get as much beta testing.  But I don't have a good feel for how many problems get resolved with Python beta releases.

**Comment by alexeypa** [comment]
> You might consider something similar with a parse_faster_args

Hm, this is an interesting suggestion...

`_parse_known_args()` is 200+ lines long. It would not be wise just clone it and apply this patch to the new version. We can probably refactor `_parse_known_args()` to make it more manageable - move sub-functions out, maybe break it up into smaller functions. It should be possible to do it and demonstrate that all these changes are just moving existing code around.

After this refactoring re-implementing the core parsing loop and exposing it as a new top-level parsing function should be relatively simple. Having two parallel functions also allow cross-checking one against the other.

I'm not sure this plan has lower risk of introducing a backward compatibility issue. The refactoring alone is going to be pretty substantial in terms of the size of the patch...

**Comment by alexeypa** [comment]
Closing this PR due to lack of interest.

**Pull request closed without merging**
